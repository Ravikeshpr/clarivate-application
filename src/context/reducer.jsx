const reducer = (state, action) => {
    switch (action.type) {
        case "ADD_DATA":
            return {
                ...state,
                list: [
                    ...new Set(
                        [...state.list, ...action.payload].map(JSON.stringify)
                    ),
                ].map(JSON.parse), //Always Unique values
            };
        case "ADD_FAVOURITE_ITEM":
            return {
                ...state,
                list: [
                    ...new Set(
                        [
                            ...state.list.toSpliced(
                                state.list.findIndex(
                                    (el) => el.id === action.payload.id
                                ),
                                1,
                                action.payload
                            ),
                        ].map(JSON.stringify)
                    ),
                ].map(JSON.parse),
            };
        case "SET_OFFSET":
            return {
                ...state,
                offset: action.payload,
            };
        case "SET_IS_API_LOAD_REQ":
            return {
                ...state,
                isAPILoadReq: action.payload,
            };
        case "SET_SCROLL_POSITION":
            return {
                ...state,
                scrollYPosition: action.payload,
            };
        default:
            return state;
    }
};

export default reducer;
