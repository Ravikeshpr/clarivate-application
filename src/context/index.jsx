/* eslint-disable react/prop-types */
import { useContext } from "react";
import { createContext, useReducer } from "react";
import reducer from "./reducer";

export const useGlobalContext = () => {
    const { state, dispatch } = useContext(globalContext);
    return { state, dispatch };
};

// Initial State
const initialState = {
    scrollYPosition: 0,
    list: [],
    offset: 1,
    isAPILoadReq: true,
};

// Create Our context
const globalContext = createContext({
    state: initialState,
    dispatch: () => {},
});

// Provider to wrap around our root react component
export const GlobalContextProvider = ({ children }) => {
    const [state, dispatch] = useReducer(reducer, initialState);
    return (
        <globalContext.Provider
            value={{
                state,
                dispatch,
            }}
        >
            {children}
        </globalContext.Provider>
    );
};
