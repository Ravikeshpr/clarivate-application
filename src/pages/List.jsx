/* eslint-disable react-hooks/exhaustive-deps */
import { Link } from "react-router-dom";
import Loader from "../hooks/useInfiniteScroll";
import { useEffect } from "react";
import { useGlobalContext } from "../context";
import Card from "../coponents/Card/Card";
import Button from "../coponents/button/Button";

export default function List() {
    const data = Loader();
    const { state, dispatch } = useGlobalContext();
    const { scrollYPosition } = state;
    useEffect(() => {
        window.scrollTo(0, scrollYPosition);
    }, []);

    const handleAddAsFavourite = (item) => {
        const tempItem = { ...item, isFavourite: true };
        dispatch({ type: "ADD_FAVOURITE_ITEM", payload: tempItem });
    };

    return (
        <>
            <div className="header-button">
                <Link to="/">
                    <Button label="Back to dashboard" />
                </Link>
            </div>
            <div className="container-fluid">
                <div className="row text-center mt-5">
                    {data &&
                        data.length > 0 &&
                        data.map((item) => {
                            return (
                                <Card
                                    key={item.id}
                                    item={item}
                                    handleAddAsFavourite={handleAddAsFavourite}
                                    type="list"
                                />
                            );
                        })}
                </div>
            </div>
        </>
    );
}
