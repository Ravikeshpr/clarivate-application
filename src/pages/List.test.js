import { BrowserRouter } from "react-router-dom";
import renderer from "react-test-renderer";
import List from "./List";
import * as GlobalContex from "../context";

describe("Dashboard", () => {
    test("renders correct", () => {
        const contextValues = {
            state: {
                scrollYPosition: 0,
                list: [
                    { id: "1", title: "title", url: "url", isFavourite: true },
                ],
                offset: 1,
                isAPILoadReq: true,
            },
            dispatch: jest.fn(),
        };

        jest.spyOn(GlobalContex, "useGlobalContext").mockImplementation(
            () => contextValues
        );
        const list = renderer.create(
            <BrowserRouter>
                <List />
            </BrowserRouter>
        );
        expect(list).toMatchSnapshot();
    });
});
