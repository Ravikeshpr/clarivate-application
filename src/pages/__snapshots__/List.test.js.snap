// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard renders correct 1`] = `
[
  <div
    className="header-button"
  >
    <a
      href="/"
      onClick={[Function]}
    >
      <button>
        Back to dashboard
      </button>
    </a>
  </div>,
  <div
    className="container-fluid"
  >
    <div
      className="row text-center mt-5"
    >
      <div
        className="card"
      >
        <img
          alt="title"
          className="card-img"
          src="url"
        />
        <div
          className="title"
        >
          <span>
            1
            :
          </span>
          <span>
             title
          </span>
        </div>
        <div>
          <div>
            <div
              className="favourite"
            >
              Marked Favourite
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
